plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Simplifier.hs

```diff
@@ -0,0 +1,96 @@
+{-# LANGUAGE NamedFieldPuns #-}
+
+module UntypedPlutusCore.Transform.Simplifier (
+    SimplifierT (..),
+    SimplifierTrace (..),
+    SimplifierStage (..),
+    Simplification (..),
+    runSimplifierT,
+    evalSimplifierT,
+    execSimplifierT,
+    Simplifier,
+    runSimplifier,
+    evalSimplifier,
+    execSimplifier,
+    initSimplifierTrace,
+    recordSimplification,
+) where
+
+import Control.Monad.State (MonadTrans, StateT)
+import Control.Monad.State qualified as State
```

Comment:
Please always either use `Control.Monad.State.Strict` or explain in a comment why the lazy version is appropriate.

Reply:
@effectfully Good point, I forgot to discuss this.

The reason I use the "lazy" version here (see below why "lazy" is in quotes) is that https://github.com/IntersectMBO/plutus/blob/bb33f082d26f8b6576d3f0d423be53eddfb6abd8/plutus-core/plutus-core/src/PlutusCore/Quote.hs#L89 is the one from the lazy version, and I didn't want to mess with it. However, notice that when updating the state I'm using the strict `modify'`.

The strange thing is that the [lazy StateT](https://hackage.haskell.org/package/transformers-0.6.1.2/docs/src/Control.Monad.Trans.State.Lazy.html#StateT) and the [strict StateT](https://hackage.haskell.org/package/transformers-0.6.1.2/docs/src/Control.Monad.Trans.State.Strict.html#StateT) have the exact same definition! There's no language strict pragma at the top, `(,)` is the regular (lazy) pair type, so unless there's something else I'm not seeing it seems like the types themselves are the same, and the only strictness/laziness changes are in the functions. So the `blabla.Strict` type is actually lazy! 😄 

Do you know if there was any intention to use the "lazy" version of `StateT` for the `MonadQuote` instance? And second of all, should we do anything about this given the above?

Reply:
> The strange thing is that the [lazy StateT](https://hackage.haskell.org/package/transformers-0.6.1.2/docs/src/Control.Monad.Trans.State.Lazy.html#StateT) and the [strict StateT](https://hackage.haskell.org/package/transformers-0.6.1.2/docs/src/Control.Monad.Trans.State.Strict.html#StateT) have the exact same definition!

That's a common thing, it's the same with `transformers` for example.

> and the only strictness/laziness changes are in the functions

Yes, but one of those functions is `>>=` and the lazy state leaks space horribly unlike the strict version.

> Do you know if there was any intention to use the "lazy" version of `StateT` for the `MonadQuote` instance?

https://getyarn.io/yarn-clip/f237528f-efda-453a-a007-8f8a7e75746d

> And second of all, should we do anything about this given the above?

Yes, let's merge your PR as it is and then try to replace all lazy state with strict state at ones. There's a chance that'll break something, but that's somewhat unlikely.